microsoft/checkedc-clang
# Introduce temporary bindings to BoundsCastExprs

This PR adds a temporary binding expression when building a BoundsCastExpr. This helps enable dynamic and assume bounds casts expressions to be treated as non-value-preserving in CanonBounds::CompareExpr (see issue #691).

### Changes
* Introduce temporary binding expression to bounds casts.
* Adjust bounds inference and visiting cast expressions to reflect the temporary bindings.
* Account for BoundsCastExprs when getting the temporary binding of an expression.
* Account for whether the expression is in a checked scope or not in `PruneTemporaryBindings`. (A follow-up issue #698 is opened to use the `CheckedScopeSpecifier` enum rather than a boolean here.)
* ~~Consider BoundsCastExprs to be value-preserving in CheckBoundsDeclarations::EqualValue. Note: this assumes that the expressions E1 and E2 have been successfully evaluated.~~ This change has been removed. This affects the Checked C static_check_bounds_cast.c test (modified in microsoft/checkedc#385 - opened #695 to track remaining work).
* Treat bounds cast expressions as non-value-preserving.
* Consider an expression e to be lexicographically equal to the usage of a temporary binding of e.
* Remove the llvm_unreachable call if a dynamic cast fails in CanonBounds::Compare. Since bounds casts are no longer considered value-preserving, it is possible to compare, e.g. a BoundsCastExpr and an ImplicitCastExpr. An ImplicitCastExpr cannot be dynamically cast to a BoundsCastExpr.
* Account for the cast kind when profiling a BoundsCastExpr.

### Tests
* Remove the function call test from bounds-decl-challenges. With the introduction of temporaries to bounds cast expressions, this test no longer produces an error.
* Add AST tests to verify that bounds cast expressions include a temporary binding expression.
* Other tests are added to checked-c (microsoft/checkedc#385).
